tinker_access_client/tinker_access_client/LcdApi.py

```diff
@@ -1,149 +1,222 @@
+      self.bus.write_byte_data(self.addr, cmd, data)
+      sleep(0.0001)
+
+# Write a block of data to the i2c bus
+   def write_block_data(self, cmd, data):
+      self.bus.write_block_data(self.addr, cmd, data)
+      sleep(0.0001)
+
+# Read a single byte from the i2c bus
+   def read(self):
+      return self.bus.read_byte(self.addr)
+
+# Read from the i2c bus
+   def read_data(self, cmd):
+      return self.bus.read_byte_data(self.addr, cmd)
+
+# Read a block of data from the i2c bus
+   def read_block_data(self, cmd):
+      return self.bus.read_block_data(self.addr, cmd)
+
+
+# LCD Address
+LCD_ADDRESS = 0x27
+
+# LCD display init sleep delay
+LCD_INITDELAY = 0.2
+
+# LCD commands
+LCD_CLEARDISPLAY = 0x01
+LCD_RETURNHOME = 0x02
+LCD_ENTRYMODESET = 0x04
+LCD_DISPLAYCONTROL = 0x08
+LCD_CURSORSHIFT = 0x10
+LCD_FUNCTIONSET = 0x20
+LCD_SETCGRAMADDR = 0x40
+LCD_SETDDRAMADDR = 0x80
+
+# flags for LCD display entry mode
+LCD_ENTRYRIGHT = 0x00
+LCD_ENTRYLEFT = 0x02
+LCD_ENTRYSHIFTINCREMENT = 0x01
+LCD_ENTRYSHIFTDECREMENT = 0x00
+
+# flags for LCD display on/off control
+LCD_DISPLAYON = 0x04
+LCD_DISPLAYOFF = 0x00
+LCD_CURSORON = 0x02
+LCD_CURSOROFF = 0x00
+LCD_BLINKON = 0x01
+LCD_BLINKOFF = 0x00
+
+# flags for LCD display/cursor shift
+LCD_DISPLAYMOVE = 0x08
+LCD_CURSORMOVE = 0x00
+LCD_MOVERIGHT = 0x04
+LCD_MOVELEFT = 0x00
+
+# flags for LCD function set
+LCD_8BITMODE = 0x10
+LCD_4BITMODE = 0x00
+LCD_2LINE = 0x08
+LCD_1LINE = 0x00
+LCD_5x10DOTS = 0x04
+LCD_5x8DOTS = 0x00
+
+# flags for LCD backlight control
+LCD_BACKLIGHT = 0x08
+LCD_NOBACKLIGHT = 0x00
+
+En = 0b00000100 # LCD Enable bit
+Rw = 0b00000010 # LCD Read/Write bit
+Rs = 0b00000001 # LCD Register select bit
 
 
 class LcdApi(object):
 
-    def __init__(self, gpio):
+    # Initialize objects and LCD
+    def __init__(self):
         self.__logger = logging.getLogger(__name__)
 
-        # noinspection PyPep8Naming
-        GPIO = self.__GPIO = gpio
-        GPIO.setup(LCD_E, GPIO.OUT)  # E
-        GPIO.setup(LCD_RS, GPIO.OUT)  # RS
-        GPIO.setup(LCD_D4, GPIO.OUT)  # DB4
-        GPIO.setup(LCD_D5, GPIO.OUT)  # DB5
-        GPIO.setup(LCD_D6, GPIO.OUT)  # DB6
-        GPIO.setup(LCD_D7, GPIO.OUT)  # DB7
+        self.lcd_device = i2c_device(LCD_ADDRESS)
+
+        self.lcd_write(0x03)
+        self.lcd_write(0x03)
+        self.lcd_write(0x03)
+        self.lcd_write(0x02)
 
         self.__init__display()
 
+    # Initialize display
     def __init__display(self):
-        # # Initialise display
-        self.lcd_byte(0x33, LCD_CMD)  # 110011 Initialise
-        self.lcd_byte(0x32, LCD_CMD)  # 110010 Initialise
-        self.lcd_byte(0x06, LCD_CMD)  # 000110 Cursor move direction
-        self.lcd_byte(0x0C, LCD_CMD)  # 001100 Display On,Cursor Off, Blink Off
-        self.lcd_byte(0x28, LCD_CMD)  # 101000 Data length, number of lines, font size
-        self.lcd_byte(0x01, LCD_CMD)  # 000001 Clear display
-        time.sleep(E_DELAY)
-
-    def __enter__(self):
```

Comment:
@kso These are are required for python context management. See [this explanation](https://jeffknupp.com/blog/2016/03/07/python-with-context-managers/)
Because you have created a new class (ic12-device) and are instantiating that the constructor for LcdApi, and deleted these lines, the context management might not work once you get past the problem with the serial port.

you will need to add these back in, and in the exit phase ensure that the new class is disposed of properly. It would probably just be easier to combine the two classes (i.e. move the methods and other code into the LcdApi  class itself so they are the same thing.